service/controller/resource/masters/create_master_instances_upgrading.go

```diff
@@ -0,0 +1,449 @@
+	"github.com/giantswarm/microerror"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/giantswarm/azure-operator/service/controller/internal/state"
+	"github.com/giantswarm/azure-operator/service/controller/key"
+)
+
+func (r *Resource) masterInstancesUpgradingTransition(ctx context.Context, obj interface{}, currentState state.State) (state.State, error) {
+	cr, err := key.ToCustomResource(obj)
+	if err != nil {
+		return "", microerror.Mask(err)
+	}
+
+	versionValue := map[string]string{}
+	{
+		for _, node := range cr.Status.Cluster.Nodes {
+			versionValue[node.Name] = node.Version
+		}
+	}
+
+	var drainerConfigs []corev1alpha1.DrainerConfig
+	{
+		n := v1.NamespaceAll
+		o := metav1.ListOptions{
+			LabelSelector: fmt.Sprintf("%s=%s", key.ClusterIDLabel, key.ClusterID(cr)),
+		}
+
+		list, err := r.g8sClient.CoreV1alpha1().DrainerConfigs(n).List(o)
+		if err != nil {
+			return "", microerror.Mask(err)
+		}
+
+		drainerConfigs = list.Items
+	}
+
+	var masterUpgradeInProgress bool
+	{
+		allMasterInstances, err := r.allInstances(ctx, cr, key.MasterVMSSName)
+		if IsScaleSetNotFound(err) {
+			r.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("did not find the scale set '%s'", key.MasterVMSSName(cr)))
+		} else if err != nil {
+			return "", microerror.Mask(err)
+		} else {
+			r.logger.LogCtx(ctx, "level", "debug", "message", "processing master VMSSs")
+
+			ws, err := r.nextInstance(ctx, cr, allMasterInstances, drainerConfigs, key.MasterInstanceName, versionValue)
+			if err != nil {
+				return "", microerror.Mask(err)
+			}
+
+			err = r.updateInstance(ctx, cr, ws.InstanceToUpdate(), key.MasterVMSSName, key.MasterInstanceName)
+			if err != nil {
+				return "", microerror.Mask(err)
+			}
+			if ws.InstanceToDrain() != nil {
+				err = r.createDrainerConfig(ctx, cr, key.MasterInstanceName(cr, *ws.InstanceToDrain().InstanceID))
+				if err != nil {
+					return "", microerror.Mask(err)
+				}
+			}
+			err = r.reimageInstance(ctx, cr, ws.InstanceToReimage(), key.MasterVMSSName, key.MasterInstanceName)
+			if err != nil {
+				return "", microerror.Mask(err)
+			}
+			err = r.deleteDrainerConfig(ctx, cr, ws.InstanceToReimage(), key.MasterInstanceName, drainerConfigs)
+			if err != nil {
+				return "", microerror.Mask(err)
+			}
+
+			masterUpgradeInProgress = ws.IsWIP()
+
+			r.logger.LogCtx(ctx, "level", "debug", "message", "processed master VMSSs")
+		}
+	}
+
+	if !masterUpgradeInProgress {
+		// When masters are upgraded, proceed to workers.
+		return WaitForMastersToBecomeReady, nil
+	}
+
+	// Upgrade still in progress. Keep current state.
+	return currentState, nil
+}
+
+func (r *Resource) allInstances(ctx context.Context, customObject providerv1alpha1.AzureConfig, deploymentNameFunc func(customObject providerv1alpha1.AzureConfig) string) ([]compute.VirtualMachineScaleSetVM, error) {
```

Comment:
We should have these functions that are common for masters and workers in some common place. Not a blocker, it can be done in follow up refactoring PR.